PR: https://github.com/odoo/odoo/pull/170334

From: cac96f5d3c391ea0d52da2e3433e7eabe07e8fb0
From: jugj-odoo
Date: 2024-09-04 11:57:41

Breaking data model changes scores: del:3 + add:0, change matches:
-    hr_attendance_overtime = fields.Boolean(related='employee_company_id.hr_attendance_overtime')
-    hr_attendance_overtime = fields.Boolean(compute='_compute_hr_attendance_overtime')
-    _inherit = 'res.company'

Total Changes: 43

[IMP] hr_attendance: Add extra hours validation mecanism

task-3999208

closes odoo/odoo#170334

Related: odoo/enterprise#65342
Related: odoo/upgrade#6426
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays_attendance/models/__init__.py
+++ b/addons/hr_holidays_attendance/models/__init__.py
@@ -6,5 +6,4 @@ from . import hr_leave_allocation
 from . import hr_leave_accrual_plan_level
 from . import hr_leave_type
 from . import hr_leave
-from . import res_company
 from . import res_users

--- a/addons/hr_holidays_attendance/models/hr_leave.py
+++ b/addons/hr_holidays_attendance/models/hr_leave.py
@@ -103,7 +103,7 @@ class HRLeave(models.Model):
     def _update_leaves_overtime(self):
         employee_dates = defaultdict(set)
         for leave in self:
-            if leave.employee_id and leave.employee_company_id.hr_attendance_overtime:
+            if leave.employee_id:
                 for d in range((leave.date_to - leave.date_from).days + 1):
                     employee_dates[leave.employee_id].add(self.env['hr.attendance']._get_day_start_and_day(leave.employee_id, leave.date_from + timedelta(days=d)))
         if employee_dates:

--- a/addons/hr_holidays_attendance/models/hr_leave_allocation.py
+++ b/addons/hr_holidays_attendance/models/hr_leave_allocation.py
@@ -25,12 +25,11 @@ class HolidaysAllocation(models.Model):
     overtime_deductible = fields.Boolean(compute='_compute_overtime_deductible')
     overtime_id = fields.Many2one('hr.attendance.overtime', string='Extra Hours', groups='hr_holidays.group_hr_holidays_user')
     employee_overtime = fields.Float(related='employee_id.total_overtime', groups='base.group_user')
-    hr_attendance_overtime = fields.Boolean(related='employee_company_id.hr_attendance_overtime')
 
     @api.depends('holiday_status_id')
     def _compute_overtime_deductible(self):
         for allocation in self:
-            allocation.overtime_deductible = allocation.hr_attendance_overtime and allocation.holiday_status_id.overtime_deductible
+            allocation.overtime_deductible = allocation.holiday_status_id.overtime_deductible
 
     @api.model_create_multi
     def create(self, vals_list):

--- a/addons/hr_holidays_attendance/models/hr_leave_type.py
+++ b/addons/hr_holidays_attendance/models/hr_leave_type.py
@@ -8,7 +8,6 @@ from odoo import _, api, fields, models
 class HRLeaveType(models.Model):
     _inherit = 'hr.leave.type'
 
-    hr_attendance_overtime = fields.Boolean(compute='_compute_hr_attendance_overtime')
     overtime_deductible = fields.Boolean(
         "Deduct Extra Hours", default=False,
         help="Once a time off of this type is approved, extra hours in attendances will be deducted.")
@@ -52,12 +51,3 @@ class HRLeaveType(models.Model):
         res = super()._get_days_request(date)
         res[1]['overtime_deductible'] = self.overtime_deductible
         return res
-
-    @api.depends('company_id.hr_attendance_overtime')
-    def _compute_hr_attendance_overtime(self):
-        # If no company is linked to the time off type, use the current company's setting
-        for leave_type in self:
-            if leave_type.company_id:
-                leave_type.hr_attendance_overtime = leave_type.company_id.hr_attendance_overtime
-            else:
-                leave_type.hr_attendance_overtime = self.env.company.hr_attendance_overtime

--- a/addons/hr_holidays_attendance/models/res_company.py
+++ b/None
@@ -1,27 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, models
-
-
-class ResCompany(models.Model):
-    _inherit = 'res.company'
-
-    @api.model
-    def _check_extra_hours_time_off(self):
-        extra_hours_time_off_type = self.env.ref('hr_holidays_attendance.holiday_status_extra_hours', raise_if_not_found=False)
-        if not extra_hours_time_off_type:
-            return
-        all_companies = self.env['res.company'].sudo().search([])
-        # Unarchive time of type if the feature is enabled
-        if any(company.hr_attendance_overtime and not extra_hours_time_off_type.active for company in all_companies):
-            extra_hours_time_off_type.toggle_active()
-        # Archive time of type if the feature is disabled for all the company
-        if all(not company.hr_attendance_overtime and extra_hours_time_off_type.active for company in all_companies):
-            extra_hours_time_off_type.toggle_active()
-
-    def write(self, vals):
-        res = super().write(vals)
-        if 'hr_attendance_overtime' in vals:
-            self._check_extra_hours_time_off()
-        return res
